### PR TITLE
Ipv4 and ipv6 domains

### PR DESCRIPTION
**Pull request by Enmk**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Added:
 * IDataTypeDomain interface;
 * method DataTypeFactory::registerDataTypeDomain for registering domains;
 * DataTypeDomainWithSimpleSerialization domain base class with simple serialization/deserialization;
 * Concrete IPv4 and IPv6 domain implementations: DataTypeDomanIPv6 and DataTypeDomanIPv4;

Updated:
 * IDataType text serialization/deserialization methods;
 * IDataType implementation to use domain for text serialization/deserialization;
 * Refactored implementation of the IPv4/IPv6 functions to use formatIPv4/v6 and parseIPv4/v6 from Common/formatIPv6.h;

Tests:
 * Added test cases for IPv4 and IPv6 domains.
 * Updated IPv4/v6 functions tests to validate more cases;
 * Added performance tests for IPv4 and IPv6 related functions;

### CODE REVIEW

File: dbms/src/DataTypes/DataTypeDomainIPv4AndIPv6.cpp [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+                throw Exception("Invalid IPv4 value.", ErrorCodes::CANNOT_PARSE_DOMAIN_VALUE_FROM_STRING);
+        }
+
+        col->insert(ipv4_value);
+    }
+};
+
+} // namespace
+
+void registerDataTypeDomainIPv4AndIPv5(DataTypeFactory & factory)
```
Comment (alexey-milovidov) [inline dbms/src/DataTypes/DataTypeDomainIPv4AndIPv6.cpp]: IPv5?

File: dbms/src/DataTypes/DataTypeDomainWithSimpleSerialization.h [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+    /** Text serialization as a literal that may be inserted into a query.
+      */
+    void serializeTextQuoted(const IColumn & column, size_t row_num, WriteBuffer & ostr, const FormatSettings &) const  override;
+    void deserializeTextQuoted(IColumn & column, ReadBuffer & istr, const FormatSettings &) const  override;
+
+    /** Text serialization for the CSV format.
+      */
+    void serializeTextCSV(const IColumn & column, size_t row_num, WriteBuffer & ostr, const FormatSettings &) const override;
+    /** delimiter - the delimiter we expect when reading a string value that is not double-quoted
```
Comment (alexey-milovidov) [inline dbms/src/DataTypes/DataTypeDomainWithSimpleSerialization.h]: Obsolete comment.

File: dbms/src/DataTypes/DataTypeDomainWithSimpleSerialization.h [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+      * (the delimiter is not consumed).
+      */
+    void deserializeTextCSV(IColumn & column, ReadBuffer & istr, const FormatSettings &) const override;
+
+    /** Text serialization intended for using in JSON format.
+      * force_quoting_64bit_integers parameter forces to brace UInt64 and Int64 types into quotes.
```
Comment (alexey-milovidov) [inline dbms/src/DataTypes/DataTypeDomainWithSimpleSerialization.h]: Obsolete comment.

File: dbms/src/DataTypes/DataTypeFactory.cpp [hunk 4]
```diff
@@ -128,6 +115,48 @@ void DataTypeFactory::registerSimpleDataType(const String & name, SimpleCreator
     }, case_sensitiveness);
 }
 
+void DataTypeFactory::registerDataTypeDomain(const String & domain_name, const String & family_name, DataTypeDomainUniquePtr domain, CaseSensitiveness case_sensitiveness)
+{
+    const auto& data_type_creator = findCreatorByName(family_name);
```
Comment (alexey-milovidov) [inline dbms/src/DataTypes/DataTypeFactory.cpp]: Style.

File: dbms/src/DataTypes/DataTypeFactory.cpp [hunk 5]
```diff
@@ -128,6 +115,48 @@ void DataTypeFactory::registerSimpleDataType(const String & name, SimpleCreator
...
+    all_domains.reserve(all_domains.size() + 1);
+
+    // We can't move the pointer to the Creator since creator is copied later.
```
Comment (alexey-milovidov) [inline dbms/src/DataTypes/DataTypeFactory.cpp]: shared_ptr is Ok.

File: dbms/src/DataTypes/IDataTypeDomain.h [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+namespace DB
+{
+
+class ReadBuffer;
+class WriteBuffer;
+struct FormatSettings;
+class IColumn;
+
+class IDataTypeDomain
```
Comment (alexey-milovidov) [inline dbms/src/DataTypes/IDataTypeDomain.h]: Missing comment.

File: dbms/src/DataTypes/IDataTypeDomain.h [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+      */
+    virtual void serializeTextCSV(const IColumn & column, size_t row_num, WriteBuffer & ostr, const FormatSettings &) const = 0;
+
+    /** delimiter - the delimiter we expect when reading a string value that is not double-quoted
+      * (the delimiter is not consumed).
+      */
+    virtual void deserializeTextCSV(IColumn & column, ReadBuffer & istr, const FormatSettings &) const = 0;
+
+    /** Text serialization intended for using in JSON format.
+      * force_quoting_64bit_integers parameter forces to brace UInt64 and Int64 types into quotes.
```
Comment (alexey-milovidov) [inline dbms/src/DataTypes/IDataTypeDomain.h]: Obsolete comment.

File: dbms/src/DataTypes/IDataTypeDomain.h [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    /** Text serialization as a literal that may be inserted into a query.
+      */
+    virtual void serializeTextQuoted(const IColumn & column, size_t row_num, WriteBuffer & ostr, const FormatSettings &) const = 0;
+    virtual void deserializeTextQuoted(IColumn & column, ReadBuffer & istr, const FormatSettings &) const = 0;
+
+    /** Text serialization for the CSV format.
+      */
+    virtual void serializeTextCSV(const IColumn & column, size_t row_num, WriteBuffer & ostr, const FormatSettings &) const = 0;
+
+    /** delimiter - the delimiter we expect when reading a string value that is not double-quoted
```
Comment (alexey-milovidov) [inline dbms/src/DataTypes/IDataTypeDomain.h]: Obsolete comment.

**Comment by champtar** [comment]
Hi @Enmk @alexey-milovidov is there any plans to have an IP datatype where you can have both IPv4 and IPv6 ? Right now I'm storing all IPs as a FixedString(16) so it's way simpler when querying

File: dbms/src/DataTypes/DataTypeDomainIPv4AndIPv6.cpp (hunk 1 above)
Comment (Enmk) [reply dbms/src/DataTypes/DataTypeDomainIPv4AndIPv6.cpp]: Fixed

File: dbms/src/DataTypes/DataTypeDomainWithSimpleSerialization.h (hunk 2 above)
Comment (Enmk) [reply dbms/src/DataTypes/DataTypeDomainWithSimpleSerialization.h]: Fixed

File: dbms/src/DataTypes/DataTypeDomainWithSimpleSerialization.h (hunk 3 above)
Comment (Enmk) [reply dbms/src/DataTypes/DataTypeDomainWithSimpleSerialization.h]: Fixed

File: dbms/src/DataTypes/IDataTypeDomain.h (hunk 6 above)
Comment (Enmk) [reply dbms/src/DataTypes/IDataTypeDomain.h]: Fixed

File: dbms/src/DataTypes/IDataTypeDomain.h (hunk 7 above)
Comment (Enmk) [reply dbms/src/DataTypes/IDataTypeDomain.h]: Fixed

File: dbms/src/DataTypes/IDataTypeDomain.h (hunk 8 above)
Comment (Enmk) [reply dbms/src/DataTypes/IDataTypeDomain.h]: Fixed

File: dbms/src/DataTypes/DataTypeFactory.cpp (hunk 5 above)
Comment (Enmk) [reply dbms/src/DataTypes/DataTypeFactory.cpp]: Simplified that piece of code, `shared_ptr` is not needed anymore.

File: dbms/src/DataTypes/DataTypeFactory.cpp (hunk 4 above)
Comment (Enmk) [reply dbms/src/DataTypes/DataTypeFactory.cpp]: Fixed too

**Comment by Enmk** [comment]
Performance comparison result.

Performance was measured on my machine against base version (without these modifications) using performance tests included in the PR. Negative numbers represent degraded performance, positive - improved, see `compare.py`` for details.

Please note that I've observed +/-1% variance of the results from run to run, so I believe there numbers here are within +/-2%.


## ipv4 (ALL CASES)
Queries:  444
Stats:
        avg:    -12.1179753049
        min:    -33.0634032212
        max:    12.7247256601
        stdev:  6.32970567083


## ipv4 IPv4StringToNum
Queries:  100
Stats:
        avg:    -16.9865447285
        min:    -23.5455008443
        max:    -12.0256227406
        stdev:  2.44602672946


## ipv4 IPv4NumToString+IPv4StringToNum
Queries:  100
Stats:
        avg:    -9.63764803906
        min:    -21.3878474019
        max:    -5.97267144694
        stdev:  1.66649394699


## ipv4 IPv4NumToStringClassC+IPv4StringToNum
Queries:  100
Stats:
        avg:    -8.24220705267
        min:    -13.3396067461
        max:    -3.68402493364
        stdev:  1.42471396986


## ipv4 error
Queries:  44
Stats:
        avg:    -1.78364095218
        min:    -12.4710148976
        max:    12.7247256601
        stdev:  7.59229876574


## ipv6 (ALL CASES)
Queries:  426
Stats:
        avg:    -5.88968274932
        min:    -22.8925432598
        max:    10.2782631191
        stdev:  5.88792824151


## ipv6 IPv6StringToNum
Queries:  100
Stats:
        avg:    -15.0270332058
        min:    -22.8925432598
        max:    -10.0665889641
        stdev:  2.67577871908


## ipv6 IPv6NumToString+IPv6StringToNum
Queries:  100
Stats:
        avg:    -5.09307059004
        min:    -10.7523961077
        max:    -2.51185092654
        stdev:  1.6369094322


## ipv6 error
Queries:  26
Stats:
        avg:    -4.67112698211
        min:    -13.9460662489
        max:    4.58204584396
        stdev:  5.82377752186


## ipv6 mapped
Queries:  200
Stats:
        avg:    -1.87772585045
        min:    -18.7335566474
        max:    10.2782631191
        stdev:  2.49766966381

File: dbms/tests/queries/0_stateless/00725_ipv4_ipv6_domains.sql [hunk 9]
```diff
@@ -0,0 +1,59 @@
+DROP TABLE IF EXISTS ipv4_test;
```
Comment (s-mx) [inline dbms/tests/queries/0_stateless/00725_ipv4_ipv6_domains.sql]: In stateless tests tables should be created in `test` database.

File: dbms/src/Common/formatIPv6.cpp [hunk 10]
```diff
@@ -46,18 +48,25 @@ static void printInteger(char *& out, T value)
...
 /// print IPv4 address as %u.%u.%u.%u
-static void formatIPv4(const unsigned char * src, char *& dst, UInt8 zeroed_tail_bytes_count)
+static void doFormatIPv4(const unsigned char * src, char *& dst, UInt8 zeroed_tail_bytes_count)
 {
     const auto limit = IPV4_BINARY_LENGTH - zeroed_tail_bytes_count;
 
-    for (const auto i : ext::range(0, IPV4_BINARY_LENGTH))
+    for (const auto i : ext::range(0, limit))
     {
         UInt8 byte = (i < limit) ? src[i] : 0;
```
Comment (s-mx) [inline dbms/src/Common/formatIPv6.cpp]: useless ternary operator

Comment (Enmk) [reply dbms/src/Common/formatIPv6.cpp]: > useless ternary operator

fixed

File: dbms/tests/queries/0_stateless/00725_ipv4_ipv6_domains.sql (hunk 9 above)
Comment (Enmk) [reply dbms/tests/queries/0_stateless/00725_ipv4_ipv6_domains.sql]: fixed

File: dbms/tests/queries/0_stateless/00076_ip_coding_functions.sql [hunk 11]
```diff
@@ -6,6 +6,8 @@ select IPv4StringToNum('127.0.0.1' as p) == (0x7f000001 as n), IPv4NumToString(n
 select IPv4StringToNum(materialize('127.0.0.1') as p) == (materialize(0x7f000001) as n), IPv4NumToString(n) == p;
 select IPv4NumToString(toUInt32(0)) == '0.0.0.0';
 select IPv4NumToString(materialize(toUInt32(0))) == materialize('0.0.0.0');
+select IPv4NumToString(toUInt32(0x7f000001)) == '127.0.0.1';
+select IPv4NumToString(materialize(toUInt32(0x7f000001))) == materialize('127.0.0.1');
```
Comment (s-mx) [inline dbms/tests/queries/0_stateless/00076_ip_coding_functions.sql]: Unfortunately, this test fails on my pc. Here is the diff https://gist.github.com/s-mx/83ed11ed435525470869cf2edc33da3d

File: dbms/src/Common/formatIPv6.cpp [hunk 12]
```diff
@@ -136,4 +145,134 @@ void formatIPv6(const unsigned char * src, char *& dst, UInt8 zeroed_tail_bytes_
     *dst++ = '\0';
 }
 
+//UInt32 parseIPv4_fast(const char * src)
```
[the remark this answers is not in the document]
Comment (Enmk) [reply dbms/src/Common/formatIPv6.cpp]: Nope, forgot to remove refactoring leftovers from commit.

File: dbms/tests/queries/0_stateless/00076_ip_coding_functions.sql (hunk 11 above)
Comment (Enmk) [reply dbms/tests/queries/0_stateless/00076_ip_coding_functions.sql]: Fixed that

**Comment by The-Alchemist** [comment]
@KochetovNicolai : Out of curiosity, why is IPv6 represented using `FixedString(16)` instead of `UInt128`?  Would a PR changing it to `UInt128` be considered?

**Comment by Enmk** [comment]
> @Enmk : Out of curiosity, why is IPv6 represented using `FixedString(16)` instead of `UInt128`?

Well, the major reason is compatibility. Existing IPv6-functions work with `FixedString(16)` and hence work with IPv6 domain too, also that allows for seamless comparing of old `FixedString(16)`-IPv6 column values with `IPv6`.

**Pull request merged**